### PR TITLE
core: shuffle members when sending messages and introductions; send to admins and owners first

### PR DESCRIPTION
**Pull request by spaced4ndy**

### CODE REVIEW

File: src/Simplex/Chat/Util.hs [hunk 1]
```diff
@@ -30,3 +33,11 @@ encryptFile fromPath toPath cfArgs = do
...
 chunkSize = 65536
 {-# INLINE chunkSize #-}
+
+shuffle :: [a] -> IO [a]
+shuffle xs = do
+  randomNumbers <- mapM (\_ -> randomRIO (0, 65535) :: IO Word16) xs
+  let zipped = zip randomNumbers xs
+      sorted = sortOn fst zipped
+      shuffled = map snd sorted
+  pure shuffled
```
Comment (epoberezkin) [inline src/Simplex/Chat/Util.hs:43]:
```suggestion
  pure $ map snd . sortOn fst . zip randomNumbers xs
```

File: src/Simplex/Chat/Util.hs [hunk 2]
```diff
@@ -30,3 +33,11 @@ encryptFile fromPath toPath cfArgs = do
 chunkSize :: Num a => a
 chunkSize = 65536
 {-# INLINE chunkSize #-}
+
+shuffle :: [a] -> IO [a]
+shuffle xs = do
+  randomNumbers <- mapM (\_ -> randomRIO (0, 65535) :: IO Word16) xs
+  let zipped = zip randomNumbers xs
+      sorted = sortOn fst zipped
```
Comment (dpwiz) [inline src/Simplex/Chat/Util.hs:41]: `sortBy (comparing fst)`?

`sortOn` does extra wrapping/unwrapping for no gain as the keys are numbers already.

Comment (dpwiz) [reply src/Simplex/Chat/Util.hs:41]: why not decorate original data while generating randoms?

Comment (epoberezkin) [reply src/Simplex/Chat/Util.hs:41]: it seems the opposite: https://hackage.haskell.org/package/base-4.19.0.0/docs/Data-List.html#v:sortOn

[sortOn](https://hackage.haskell.org/package/base-4.19.0.0/docs/Data-List.html#v:sortOn) f is equivalent to [sortBy](https://hackage.haskell.org/package/base-4.19.0.0/docs/Data-List.html#v:sortBy) ([comparing](https://hackage.haskell.org/package/base-4.19.0.0/docs/Data-Ord.html#v:comparing) f), but has the performance advantage of only evaluating f once for each element in the input list.

Comment (epoberezkin) [reply src/Simplex/Chat/Util.hs:41]: I think I understand what you say now - it's better to evaluate fst twice than to move to another tuple... Interesting.

**Pull request merged**